File: python/007.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def isPrime(n, primes):
    for p in primes:
        if p * p > n:
            break
        if n % p == 0:
            return False
    return True

def p7(max):
    primes = [2, 3, 5]
    i=1
    while len(primes) < max:
        p = 6 * i + 1
        if isPrime(p, primes):
            primes.append(p)
        p += 4
        if isPrime(p, primes):
            primes.append(p)
        i+=1
    return primes.pop()

def main():
    print(p7(10001))

if __name__ == '__main__':
    main()

File: elixir/007.exs
#!/usr/bin/env elixir

# https://projecteuler.net/problem=7

defmodule P7 do
	def solve(n) do
		List.last(genPrime(n))
	end
	def genPrime(n) do
		genPrime(7, n, [2, 3, 5])
	end
	def genPrime(_, n, l) when length(l) >= n do
		l
	end
	def genPrime(i, n, l) do
		if isPrime(i, l) do
			genPrime(i+2, n, l ++ [i])
		else
			genPrime(i+2, n, l)
		end
	end
	def isPrime(i, [h|_]) when h * h > i do true end
	def isPrime(i, [h|_]) when rem(i, h) == 0 do false end
	def isPrime(i, [_|t]) do isPrime(i, t) end
	def isPrime(_, []) do true end
end

P7.solve(10_001)
|> IO.inspect
